### PR TITLE
Disallow multi variable declarations with mismatching number of values.

### PR DESCRIPTION
**Pull request by ekpyron**

### CODE REVIEW

File: test/libsolidity/syntaxTests/multiVariableDeclaration/oneElementTuple.sol [hunk 1]
```diff
@@ -0,0 +1,8 @@
+contract C {
+    function f() public {
+        (uint a,) = (1,);
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/multiVariableDeclaration/oneElementTuple.sol:3]: I'm not sure whether the right hand side even needs to be well-formed in the future (once tuple declarations and assignments with mismatching numbers of components are disallowed).

File: test/libsolidity/syntaxTests/nameAndTypeResolution/239_multi_variable_declaration_wildcards_fine.sol [hunk 2]
```diff
@@ -13,7 +13,8 @@ contract C {
...
 // ----
-// Warning: (179-198): Different number of components on the left hand side (2) than on the right hand side (3).
-// Warning: (208-233): Different number of components on the left hand side (3) than on the right hand side (2).
-// Warning: (243-262): Different number of components on the left hand side (2) than on the right hand side (3).
-// Warning: (272-297): Different number of components on the left hand side (3) than on the right hand side (2).
+// TypeError: (179-198): Different number of components on the left hand side (2) than on the right hand side (3).
+// TypeError: (208-233): Different number of components on the left hand side (3) than on the right hand side (2).
+// TypeError: (243-262): Different number of components on the left hand side (2) than on the right hand side (3).
+// TypeError: (272-297): Different number of components on the left hand side (3) than on the right hand side (2).
+// TypeError: (335-350): Variable declaration statement does not declare any variables.
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/nameAndTypeResolution/239_multi_variable_declaration_wildcards_fine.sol]: The last two refer to weird leftover ``var`` cases that should be disallowed in a separate PR.

File: test/libsolidity/syntaxTests/nameAndTypeResolution/244_tuples.sol [hunk 3]
```diff
@@ -1,13 +1,10 @@
 contract C {
-    function f() public {
+    function f() public pure {
         uint a = (1);
-        (uint b,) = (uint8(1),);
+        uint b = (uint8(1),);
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/nameAndTypeResolution/244_tuples.sol]: Not sure whether this should be valid in the future.

File: test/libsolidity/syntaxTests/nameAndTypeResolution/246_multi_variable_declaration_wildcards_fail_5.sol [hunk 4]
```diff
@@ -3,4 +3,4 @@ contract C {
     function f() public { var (,) = one(); }
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/nameAndTypeResolution/246_multi_variable_declaration_wildcards_fail_5.sol]: This should also be disallowed as a case of ``var`` in a separate PR.

**Comment by ekpyron** [comment]
~~Documentation still needs to be updated.~~

File: Changelog.md [hunk 5]
```diff
@@ -30,6 +30,7 @@ Breaking Changes:
  * Parser: Remove ``constant`` as function state mutability modifer.
  * Type Checker: Disallow arithmetic operations for boolean variables.
  * Type Checker: Disallow conversions between ``bytesX`` and ``uintY`` of different size.
+ * Type Checker: Disallow multi-variable declarations with mismatching number of values.
```
Comment (chriseth) [inline Changelog.md]: Was this an experimental 050 feature?

File: libsolidity/analysis/TypeChecker.cpp [hunk 6]
```diff
@@ -1107,80 +1107,28 @@ bool TypeChecker::visit(VariableDeclarationStatement const& _statement)
...
+	for (size_t i = 0; i < min(variables.size(), valueTypes.size()); ++i)
 	{
-		if (!assignments[i])
+		if (!variables[i])
 			continue;
-		VariableDeclaration const& var = *assignments[i];
+		VariableDeclaration const& var = *variables[i];
 		solAssert(!var.value(), "Value has to be tied to statement.");
 		TypePointer const& valueComponentType = valueTypes[i];
 		solAssert(!!valueComponentType, "");
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp:1144]: Is `!var.annotation().type` (one line further down) still possible? I guess this is out of scope for this PR.

File: libsolidity/ast/ASTAnnotations.h [hunk 7]
```diff
@@ -166,9 +166,6 @@ struct UserDefinedTypeNameAnnotation: TypeNameAnnotation
 
 struct VariableDeclarationStatementAnnotation: StatementAnnotation
 {
-	/// Information about which component of the value is assigned to which variable.
```
Comment (chriseth) [inline libsolidity/ast/ASTAnnotations.h]: Can you please delete the whole struct?

File: libsolidity/codegen/ContractCompiler.cpp [hunk 8]
```diff
@@ -821,13 +821,13 @@ bool ContractCompiler::visit(VariableDeclarationStatement const& _variableDeclar
 			valueTypes = tupleType->components();
 		else
 			valueTypes = TypePointers{expression->annotation().type};
-		auto const& assignments = _variableDeclarationStatement.annotation().assignments;
+		auto const& assignments = _variableDeclarationStatement.declarations();
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: Please rename the local variable.

File: libsolidity/codegen/ContractCompiler.cpp [hunk 9]
```diff
@@ -821,13 +821,13 @@ bool ContractCompiler::visit(VariableDeclarationStatement const& _variableDeclar
...
 		solAssert(assignments.size() == valueTypes.size(), "");
 		for (size_t i = 0; i < assignments.size(); ++i)
 		{
 			size_t j = assignments.size() - i - 1;
 			solAssert(!!valueTypes[j], "");
-			VariableDeclaration const* varDecl = assignments[j];
+			VariableDeclaration const* varDecl = assignments[j].get();
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: Perhaps
```
if (VariableDeclaration const* varDecl = assignments[j].get())
{
...
}
else
  utils.popStackElement(*valueTypes[j]);
```

File: test/libsolidity/syntaxTests/multiVariableDeclaration/oneElementTuple.sol (hunk 1 above)
Comment (chriseth) [reply test/libsolidity/syntaxTests/multiVariableDeclaration/oneElementTuple.sol:3]: What do you mean by "well-formed"?

File: test/libsolidity/syntaxTests/nameAndTypeResolution/239_multi_variable_declaration_wildcards_fine.sol (hunk 2 above)
Comment (chriseth) [reply test/libsolidity/syntaxTests/nameAndTypeResolution/239_multi_variable_declaration_wildcards_fine.sol]: The test is called "..._fine", also it mentions "wildcards", so I'm not sure if the test makes any sense at all anymore.

Comment (chriseth) [reply test/libsolidity/syntaxTests/nameAndTypeResolution/239_multi_variable_declaration_wildcards_fine.sol]: Also there are other tests with "wildcards" in their name, please modify all of them (or just delete them if we already have enough tests) so that they match what we actually want to test.

File: test/libsolidity/syntaxTests/parsing/multi_variable_declarations.sol [hunk 10]
```diff
@@ -11,5 +11,5 @@ contract C {
 	function g() pure public returns (uint, uint, uint) {}
 }
 // ----
-// Warning: (93-107): Different number of components on the left hand side (2) than on the right hand side (1).
-// Warning: (111-124): Different number of components on the left hand side (2) than on the right hand side (1).
+// TypeError: (93-107): Different number of components on the left hand side (2) than on the right hand side (1).
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/parsing/multi_variable_declarations.sol]: This is a parsing test, so it should perhaps be modified to not cause any errors during type checking.

File: test/libsolidity/syntaxTests/parsing/tuples.sol [hunk 11]
```diff
@@ -8,9 +8,5 @@ contract C {
 	}
 }
 // ----
-// Warning: (54-70): Different number of components on the left hand side (2) than on the right hand side (1).
-// Warning: (107-128): Different number of components on the left hand side (2) than on the right hand side (3).
-// Warning: (75-81): Unused local variable.
-// Warning: (83-89): Unused local variable.
-// Warning: (108-114): Unused local variable.
-// Warning: (14-143): Function state mutability can be restricted to pure
+// TypeError: (54-70): Different number of components on the left hand side (2) than on the right hand side (1).
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/parsing/tuples.sol]: Same here.

File: test/libsolidity/syntaxTests/tupleAssignments/warn_fill_vardecl.sol [hunk 12]
```diff
@@ -5,4 +5,4 @@ contract C {
 	}
 }
 // ----
-// Warning: (76-99): Different number of components on the left hand side (3) than on the right hand side (4).
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/tupleAssignments/warn_fill_vardecl.sol]: This is also a wildcard test.

**Review by chriseth — DISMISSED** [review]
Please adjust the tests so that we have tests for
 - checking that the old wildcard behaviour is gone
 - checking that the number of components is counted properly, also in edge cases (0,1,2 components, less components left, less components right)

and remove any superfluous tests.

File: test/libsolidity/syntaxTests/multiVariableDeclaration/oneElementTuple.sol (hunk 1 above)
Comment (ekpyron) [reply test/libsolidity/syntaxTests/multiVariableDeclaration/oneElementTuple.sol:3]: Syntactically well formed, so whether we should disallow ``(1,)`` syntactically at some point (as already discussed elsewhere).

File: libsolidity/ast/ASTAnnotations.h (hunk 7 above)
Comment (ekpyron) [reply libsolidity/ast/ASTAnnotations.h]: Ok - there's also ``EventDefinitionAnnotation``, ``ModifierDefinitionAnnotation`` and ``StatementAnnotation`` that are empty as well - maybe all of them should be removed...

Comment (chriseth) [reply libsolidity/ast/ASTAnnotations.h]: Yes.

File: libsolidity/analysis/TypeChecker.cpp (hunk 6 above)
Comment (ekpyron) [reply libsolidity/analysis/TypeChecker.cpp:1144]: I think it is possible in error cases, as long as ``var`` errors are not fatal.

File: libsolidity/ast/ASTAnnotations.h (hunk 7 above)
Comment (chriseth) [reply libsolidity/ast/ASTAnnotations.h]: Ah no wait a second. The others also inherit from `DocumentedAnnotation`, so they are not really "empty", but this one is identical to StatementAnnotation.

Comment (ekpyron) [reply libsolidity/ast/ASTAnnotations.h]: Ok, they are still used in DocStringAnalyzer. I'll just remove this one here then.

**Comment by chriseth** [comment]
Is there a test that does `(uint a, uint b) = f()` where `f()` returns nothing? Some for `(uint a) = f()` and `uint a = f()`?

**Review by chriseth — DISMISSED** [review]

**Comment by ekpyron** [comment]
@chriseth Now there is.

**Review by chriseth — DISMISSED** [review]

**Comment by ekpyron** [comment]
Sorry, I just rebased one last time, after #4465 was merged, to be sure there was no conflict in the tests.

**Comment by ekpyron** [comment]
Which was a good idea, since taking both PRs together another v050 variable can and has to be removed.

**Comment by chriseth** [comment]
Build fails.

**Comment by ekpyron** [comment]
Oh - I must have rebased to an not-up-to-date version of ``develop``. Should work now.

**Review by chriseth — DISMISSED** [review]
Please rebase.

**Comment by ekpyron** [comment]
@chriseth Rebased. I had to change the argument types of ``createTupleDecl`` and ``typeCanBeExpressed``, though, but it should still be fine.

**Comment by chriseth** [comment]
Please rebase.

**Comment by ekpyron** [comment]
Rebased... again :). I hope after I don't know how many non-trivial rebases, everything that might conflict with this is merged already now. And hopefully I didn't mess up anything during all the rebasing :).

**Review by chriseth — APPROVED** [review]
